feat: add generic Snakemake executor

dominikburri opened the pull request:
This PR introduces a general executor for Snakemake within the python models using [`subprocess`](https://docs.python.org/3/library/subprocess.html#module-subprocess).

The idea is to obtain a `Run` object, prepare a list of strings for the snakemake call and executing it with `subprocess`. 
The executor is general and should work on any snakefile.

Needed to add `snakemake` as new requirement.

Remaining issues:

- [x] how to properly set working directory (assuming submodules in repo)
- [x] code cleaning (see checklist)
- [ ] documentation

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

### Checklist

Please carefully read these items and tick them off if the statements are true
_or do not apply_.

- [x] I have performed a self-review of my own code
- [x] My code follows the existing coding style, lints and generates no new
      warnings
- [x] I have added type annotations to all function/method signatures, and I
      have added type annotations for any local variables that are non-trivial,
      potentially ambiguous or might otherwise benefit from explicit typing.
- [x] I have commented my code in hard-to-understand areas
- [x] I have added ["Google-style docstrings"] to all new modules, classes,
      methods/functions or updated previously existing ones
- [x] I have added tests that prove my fix is effective or that my feature
      works
- [x] New and existing unit tests pass locally with my changes and I have not
      reduced the code coverage relative to the previous state
- [ ] I have updated any sections of the app's documentation that are affected
      by the proposed changes

If for some reason you are unable to tick off all boxes, please leave a
comment explaining the issue you are facing so that we can work on it
together.

uniqueg commented on requirements.txt:
```diff
@@ -1 +1,2 @@
 pydantic==1.8.1
+snakemake==7.2.1
```
The question is whether that is going to be required if the ZARP pipeline is installed anyway in #24. But okay to keep, just something to consider

dominikburri replied:
In commit e2eeea59691b15332e82dd4d7cdeb440298fef98, I removed the requirement, but now the tests won't pass anymore here. For me locally they pass because I still have `snakemake` within miniconda.

What is the best way to deal with this?

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class SnakemakeExecutor:
+    """Run snakemake with system calls."""
```
Please write a [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstring at least with the following sections:
- `Args`
- `Attributes`

`Args` refers to the arguments for the constructor (here: `run`), `Attributes` refers to the attributes of the class (here: `run_dict` and `success`).

Also nice to include a section `Examples`.

Here's an example for a class docstring with `Args`, `Attributes` and `Examples` (not a particularly good class, but good enough as an example for a docstring): https://github.com/zavolanlab/htsinfer/blob/bdccede0c3948e591f77f9a4112e7fef768dd1b5/htsinfer/get_library_type.py#L26-L56

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -0,0 +1,49 @@
+        self.run_dict = run.dict()
+        self.success = None
+
+    def prepare_run(self, snkfile: str, workdir: str) -> None:
```
All methods should also have complete [Google-style docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html), i.e., they should have sections `Args`, `Returns` and `Raises` as applicable.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -0,0 +1,49 @@
+        if self.run_dict['execution_profile'] == "local-conda":
+            prof = ["--profile", os.path.join("submodules", "zarp", "profiles", "local-conda")]
+            run_list.extend(prof)
+        self.run_list = run_list
```
This attribute is not defined in the constructor. Also, it's not good practice to modify class attributes silently within methods - it's the equivalent of modifying global variables in functions ("side effects"). Best is to define an empty run list in the constructor, and have this method return `run_list`, then assign it to `self.run_list` in the calling context.

Remember: be careful when assigning to `self.some_attribute` in a method - or your method will have side effects! I only really became conscious of that recently myself during a hackathon, and HTSinfer, for example, is full of this stuff - but it's already biting my back because it makes the code quite fragile. The exception to this anti-pattern are setter methods by the way - but they are deliberately short and explicit, so that it reduces the risks of inadvertently suffering from side effects.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -0,0 +1,49 @@
+        # Snakemake config
+        run_list.extend(["--snakefile", snkfile])
+        run_list.extend(["--cores", str(self.run_dict["cores"])])
+        # run_list.extend(["--config", f"workdir={workdir}"])
```
Make sure to remove or uncomment such comments in your final PR

dominikburri replied:
Right now, no workdir is set, so I possibly still need this, but since I did not test it, it's uncommented.
But sure, I try to remove all spurious code for the final PR.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -0,0 +1,49 @@
+        # run_list.extend(["--config", f"workdir={workdir}"])
+        # execution profile
+        if self.run_dict['execution_profile'] == "local-conda":
+            prof = ["--profile", os.path.join("submodules", "zarp", "profiles", "local-conda")]
```
I'd recommend to use the excellent `pathlib` module for any path operations in Python3. It's really much better than the older `os.path` module.

dominikburri replied:
I'll have a look into this, but not my priority right now :-)

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -0,0 +1,49 @@
+                print(f"Snakemake not finished: {completed}")
+                self.success = False
+        except subprocess.CalledProcessError as e:
+            print(e)
```
Note that you are just printing out the error message here, then ZARP-cli continues. It's probably better to raise this error. And you may wanna keep track of STDERR and print _that_. Similar to here: https://github.com/zavolanlab/htsinfer/blob/cf162d5ff8b7b2988d2bd4fb0c799a032de3ce5b/htsinfer/get_read_orientation.py#L208-L215

Although it _is_ a good idea to wrap the code in a `try` - `catch` for `CalledProcessError` as you do.

So I would:
- check the return code and print STDERR if it's non-zero
- catch `CalledProcessError`
- in both caises raise the error to the calling context (don't forget to document that in the `Raises` section in the docstring!)

Note that you can reraise exceptions in Python3. The pattern is:

```
except SomeException as exc:
    raise SameExceptionOrAnotherOne("Some descriptive error message") from exc
```

So the interesting part here is the `from`. This way you will keep all information about the original error but _also_ have the ability to add your own stuff to the error message/traceback.

dominikburri replied:
the variable `completed` contains all information, including STDERR. 
So one LOGGING is included, that could be passed. But not sure if it should raise an error. At least, if it is, the attribute `success` could be removed.

dominikburri replied:
Ok, I changed it such that [`subprocess`](https://docs.python.org/3/library/subprocess.html#module-subprocess) will raise a `CalledProcessError` when exit code != 0. I think this it the intended way. 
I still set the attribute `success` but I don't think it has a real use case here. Anyway, I'd like to keep it for now, to not loose myself in details :-).

uniqueg commented on tests/run/test_snakemake.py:
```diff
@@ -0,0 +1,63 @@
+"""Tests for module `zarp.run.snakemake"""
```
Tests look good - however, _all_ of them are integration tests. You would also need to provide _unit_ tests, where you test each individual method/function in isolation

dominikburri replied:
I added `test_run_subprocess` which I think is a unit test. Please let me know if this is what you meant.

dominikburri commented:
Now I get an error in _TravisCI_ because mamba is not installed and snakemake throws an error. Could either be fixed by 

- adding `--conda-frontend conda` (instead of the default `mamba`), see [here](https://snakemake.readthedocs.io/en/v7.2.1/executing/cli.html#CONDA). 
- or having `mamba` in the virtual environment.

Will we switch to having mamba @uniqueg ? Really strange that snakemake throws an error here. 
I could also remove the tests that actually execute snakemake, as it's maybe too much. I'm open for suggestions!

dominikburri commented:
One possible solution for the virtual environments (with required packages) is to pass hand over an environment variables when calling `subprocess`. From the [documentation](https://docs.python.org/3/library/subprocess.html):
> If env is not None, it must be a mapping that defines the environment variables for the new process; these are used instead of the default behavior of inheriting the current process’ environment.

The environment can be independently created and then referenced. 

Not totally sure if this is the way to go, but others had similar problems/requests, see discussions [here](https://stackoverflow.com/questions/2231227/python-subprocess-popen-with-a-modified-environment) and [here](https://stackoverflow.com/questions/8052926/running-subprocess-within-different-virtualenv-with-python).
I think this is a good starting point to properly use `subprocess` with various environments.

dominikburri commented:
I updated the branch with the latest dev and adjusted the tests to work again.
The only test failing is `test_execution_profile`, because I didn't find it in the model or enums. 
Will it be included or removed altogether?

uniqueg commented:
Awesome, thanks a lot. I'll have a look about this and try to merge this in this week. I'll take care of the test, too.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -1 +1,126 @@
+        self.success: Optional[bool] = None
+        self.run_list: List[str] = []
+
+    def set_run_list(self, run_list: list) -> None:
```
Here and elsewhere:

Getters and setters aren't really used in Python. And if people insist on using them, then the best way to use them is through properties. See here: https://stackoverflow.com/questions/2627002/whats-the-pythonic-way-to-use-getters-and-setters

The important part for "setter" methods is that they do not return anything, i.e., no method should set an attribute AND return something. Or, for that matter, set more than one attribute. I will rename the `prepare_run` method to `set_command` to make this clearer.

I will remove the getters and setters in the next commit.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -1 +1,126 @@
+    def __init__(self, run: InitRun) -> None:
+        """Class constructor."""
+        self.run_dict = run.dict()
+        self.success: Optional[bool] = None
```
`Optional[bool]` seems weird to me. I will add an enumerator with states instead, as it is more extensible. For now, the states will be `UNKNOWN`, `SUCCESS` and `ERROR`.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -1 +1,126 @@
+        """Class constructor."""
+        self.run_dict = run.dict()
+        self.success: Optional[bool] = None
+        self.run_list: List[str] = []
```
Will rename this to `self.command`

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -1 +1,126 @@
+            run_list.extend(["--directory",
+                            str(self.run_dict["working_directory"])])
+        # execution profile
+        if "execution_profile" in self.run_dict:
```
As this is not defined at the moment, I don't think we need to test for this now. Once we see if and how we support profiles, we can introduce this again, but for now I will remove this check.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -1 +1,126 @@
+
+    def __init__(self, run: InitRun) -> None:
+        """Class constructor."""
+        self.run_dict = run.dict()
```
I'm not quite sure why you converted the model to a dictionary. It's much nicer to work with attributes: accessing them is shorter and easier to read and write compared to accessing dictionary values. Also, editors will be aware of them semantically, which can help prevent mistakes. It's a big part of the reason why we use models in the first place.

So I will change this to use the model instead.

uniqueg commented on zarp/run/snakemake.py:
```diff
@@ -1 +1,126 @@
+            run_list.extend(conf)
+        self.set_run_list(run_list)
+
+    def validate_run(self) -> bool:
```
It needs more than just `snakemake` for a valid command. What exactly is needed isn't even clear. Let Snakemake do the the validation - it will surely complain (and complain with a more helpful/explicit error message than we would). Will remove this.

uniqueg reviewed (changes requested):
Thanks a lot @dominikburri. As I said, I will take it from here. Just gave some feedback on things that I will change (in fact, I have already changed them, just haven't pushed yet).
